### PR TITLE
gh-91621: Fix typing.get_type_hints for collections.abc.Callable

### PR DESCRIPTION
**Pull request by hauntsaninja**
This mirrors logic in typing.get_args. The trickiness comes from how we
flatten args in collections.abc.Callable, see
https://bugs.python.org/issue42195

#91621

### CODE REVIEW

**Comment by hauntsaninja** [comment]
I think we can skip news since this is just a fix for the unreleased https://github.com/python/cpython/pull/30900

**Comment by AlexWaygood** [comment]
> I think we can skip news since this is just a fix for the unreleased #30900

Was #30900 not included in 3.11a7? If it was, then I think this probably does need a short NEWS entry :)

File: Lib/test/test_typing.py [hunk 1]
```diff
@@ -4408,6 +4408,12 @@ def test_get_type_hints_typeddict(self):
             'a': Annotated[Required[int], "a", "b", "c"]
         })
 
+    def test_get_type_hints_collections_abc_callable(self):
+        # https://github.com/python/cpython/issues/91621
+        def f(x: collections.abc.Callable[[int], int]): ...
+
+        self.assertEqual(get_type_hints(f), {'x': collections.abc.Callable[[int], int]})
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py:4418]: Can you add tests for some edge cases like `Callable[..., int]` and `Callable[P, int]`?

File: Lib/typing.py [hunk 2]
```diff
@@ -350,7 +350,11 @@ def _eval_type(t, globalns, localns, recursive_guard=frozenset()):
                 ForwardRef(arg) if isinstance(arg, str) else arg
                 for arg in t.__args__
             )
-            t = t.__origin__[args]
+            if (t.__origin__ is collections.abc.Callable
```
Comment (Fidget-Spinner) [inline Lib/typing.py]: I know this code pattern appears once in Generic code and once here now. Should we just convert this into one function? Maybe `_should_flatten_callable_args()`? Seems abit tiresome to repeat this (and it's also confusing for people who haven't dealt with Callable's args before).

**Review by Fidget-Spinner — COMMENTED** [review]
LGTM

File: Lib/typing.py (hunk 2 above)
Comment (hauntsaninja) [reply Lib/typing.py]: (sorry, missed this comment)
I added a helper function for this

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**